tapdb/multiverse.go

```diff
@@ -61,21 +70,238 @@ type BatchedMultiverse interface {
 	BatchedTx[BaseMultiverseStore]
 }
 
+// ProofKey is used to uniquely identify a proof within a universe. This is
+// used for the LRU cache for the proofs themselves, which are considered to be
+// immutable.
+type ProofKey [32]byte
+
+// NewProofKey takes a universe identifier and leaf key, and returns a proof
+// key.
+func NewProofKey(id universe.Identifier, key universe.LeafKey) ProofKey {
+	idBytes := id.Bytes()
+	leafKeyBytes := key.UniverseKey()
+
+	// The proof key maps down the ID and the leaf key into a single
+	// 32-byte value: sha256(id || leaf_key)..
+	h := sha256.New()
+	h.Write(idBytes[:])
+	h.Write(leafKeyBytes[:])
+
+	return fn.ToArray[ProofKey](h.Sum(nil))
+}
+
+// numCachedProofs is the number of universe proofs we'll cache.
+const numCachedProofs = 25_000
+
+// cachedProof is a single cached proof.
+type cachedProof []*universe.Proof
+
+// Size just returns 1 as we're limiting based on the total number of proofs.
+func (c *cachedProof) Size() (uint64, error) {
+	return 1, nil
+}
+
+// leafProofCache is used to cache proofs for issuance leaves for assets w/o a
+// group key.
+type leafProofCache = *lru.Cache[ProofKey, *cachedProof]
```

Comment:
IMO it's quite hard to reason about type aliases that are referencing a pointer type. Should we just do `type leafProofCache = *lru.Cache[ProofKey, *cachedProof]` and then return a pointer of `leafProofCache` below?

Reply:
This would also better match the pattern of the `proofCache` below.

Reply:
This is needed to have some of the declarations compile cleanly. Otherwise it needs to do some intermediate caching.